File: python/src/fgraph/values.py
"""Canonical value encoding shared by storage, query, portable events, and CLI surfaces."""

from __future__ import annotations

import base64
import binascii
import builtins
import hashlib
import json
import math
import re
import struct
from collections.abc import Callable, Mapping, Sequence
from dataclasses import dataclass
from datetime import UTC, datetime, timedelta
from typing import Any

from fgraph.errors import TooLarge
from fgraph.errors import TypeError as FGraphTypeError

BLOB_THRESHOLD = 256
MAX_VALUE_BYTES = 1_048_576
MAX_JSON_DEPTH = 64
MAX_JSON_DOCUMENT_DEPTH = 80
INT64_MIN = -(2**63)
INT64_MAX = 2**63 - 1
INSTANT_MIN = -62_135_596_800_000_000
INSTANT_MAX = 253_402_300_799_999_999
_UNIX_EPOCH = datetime(1970, 1, 1, tzinfo=UTC)
RFC3339_PATTERN = re.compile(
    r"[0-9]{4}-[0-9]{2}-[0-9]{2}T[0-9]{2}:[0-9]{2}:[0-9]{2}(?:\.[0-9]+)?(?:Z|[+-][0-9]{2}:[0-9]{2})"
)
ATTRIBUTE_PATTERN = re.compile(r"^[a-z0-9][a-z0-9._-]*/[a-z0-9][a-z0-9._-]*$")

REF = 0
BOOL = 1
INT = 2
FLOAT = 3
TEXT = 4
INSTANT = 5
BYTES = 6
VECTOR = 7
TEXT_REF = 8
BYTES_REF = 9
JSON = 10

TAG_NAMES = {
    REF: "ref",
    BOOL: "bool",
    INT: "int",
    FLOAT: "float",
    TEXT: "text",
    INSTANT: "instant",
    BYTES: "bytes",
    VECTOR: "vector",
    TEXT_REF: "text",
    BYTES_REF: "bytes",
    JSON: "json",
}
TYPE_NAMES = frozenset({"ref", "bool", "int", "float", "text", "instant", "bytes", "vector", "json"})


def indirect_digest(tag: int, data: bytes) -> bytes:
    """Hash indirect content with its physical tag as a one-byte domain."""
    return hashlib.sha256(bytes((tag,)) + data).digest()


@dataclass(frozen=True, slots=True)
class Encoded:
    """A logical value reduced to its normative SQLite representation."""

    tag: int
    stored: int | float | str | bytes
    logical: Any
    blob: str | bytes | None = None


@dataclass(frozen=True, slots=True)
class Cell:
    """Hashable typed value used by the in-process Datalog evaluator."""

    tag: int
    value: Any


def canonical_json(value: Any) -> str:
    """Encode JSON with fgraph's canonical key, string, and binary64 number rules."""
    return _canonical_json(value, max_depth=MAX_JSON_DEPTH)


def _canonical_json_document(value: Any) -> str:
    """Encode a validated wire envelope without charging it to a logical value's depth."""
    return _canonical_json(value, max_depth=MAX_JSON_DOCUMENT_DEPTH)


def _canonical_json(value: Any, *, max_depth: int) -> str:
    try:
        result = _canonical_part(value, depth=0, max_depth=max_depth)
    except (ValueError, builtins.TypeError) as exc:
        raise FGraphTypeError(f"invalid JSON value {value!r}; use finite JSON scalars, arrays, and objects") from exc
    return result


def _canonical_float(value: float) -> str:
    if not math.isfinite(value):
        raise ValueError("non-finite JSON number")
    if value == 0:
        return "0"
    if value.is_integer() and INT64_MIN <= int(value) <= INT64_MAX:
        # Preserve the exact integer represented by binary64 at the int64
        # boundary instead of expanding the rounded repr() mantissa.
        return str(int(value))
    rendered = repr(value).lower()
    absolute = abs(value)
    if "e" not in rendered:
        return rendered.removesuffix(".0")
    mantissa, exponent_text = rendered.split("e")
    exponent = int(exponent_text)
    integral_outside_int64 = value.is_integer()
    if 1e-6 <= absolute < 1e21 and not integral_outside_int64:
        sign = ""
        if mantissa.startswith("-"):
            sign, mantissa = "-", mantissa[1:]
        digits = mantissa.replace(".", "")
        decimal_position = 1 + exponent
        if decimal_position <= 0:
            return f"{sign}0.{('0' * -decimal_position)}{digits}"
        if decimal_position >= len(digits):
            return f"{sign}{digits}{'0' * (decimal_position - len(digits))}"
        return f"{sign}{digits[:decimal_position]}.{digits[decimal_position:]}"
    normalized_mantissa = mantissa.removesuffix(".0")
    exponent_sign = "+" if exponent >= 0 else "-"
    return f"{normalized_mantissa}e{exponent_sign}{abs(exponent)}"


def _canonical_part(value: Any, *, depth: int, max_depth: int) -> str:
    if value is None:
        return "null"
    if value is True:
        return "true"
    if value is False:
        return "false"
    if isinstance(value, int):
        if not INT64_MIN <= value <= INT64_MAX:
            raise ValueError("JSON integer exceeds signed 64-bit range")
        return str(value)
    if isinstance(value, float):
        return _canonical_float(value)
    if isinstance(value, str):
        _utf8(value)
        return json.encoder.encode_basestring(value)
    if isinstance(value, Mapping):
        child_depth = depth + 1
        if child_depth > max_depth:
            raise TooLarge(f"JSON nesting depth exceeds {max_depth}; flatten deeply nested arrays and objects")
        if not all(isinstance(key, str) for key in value):
            raise builtins.TypeError("JSON object keys must be strings")
        for key in value:
            _utf8(key)
        return (
            "{"
            + ",".join(
                f"{json.encoder.encode_basestring(key)}:"
                f"{_canonical_part(value[key], depth=child_depth, max_depth=max_depth)}"
                for key in sorted(value)
            )
            + "}"
        )
    if isinstance(value, Sequence) and not isinstance(value, (str, bytes, bytearray)):
        child_depth = depth + 1
        if child_depth > max_depth:
            raise TooLarge(f"JSON nesting depth exceeds {max_depth}; flatten deeply nested arrays and objects")
        return "[" + ",".join(_canonical_part(item, depth=child_depth, max_depth=max_depth) for item in value) + "]"
    raise builtins.TypeError(f"unsupported JSON value {value!r}")


def _check_size(data: bytes, value: Any) -> None:
    if len(data) > MAX_VALUE_BYTES:
        raise TooLarge(
            f"value {type(value).__name__} is {len(data)} bytes; keep fact values at or below {MAX_VALUE_BYTES} bytes"
        )


def _utf8(value: str) -> bytes:
    try:
        return value.encode()
    except UnicodeEncodeError as exc:
        raise FGraphTypeError(f"string {value!r} is not valid UTF-8; remove unpaired surrogate code points") from exc


def _instant(value: Any) -> int:
    if isinstance(value, bool):
        raise FGraphTypeError(f"invalid instant {value!r}; use integer microseconds or an RFC 3339 UTC string")
    if isinstance(value, int):
        if INSTANT_MIN <= value <= INSTANT_MAX:
            return value
        raise FGraphTypeError(
            f"instant {value!r} is outside RFC 3339 years 0001..9999; use representable UTC microseconds"
        )
    if isinstance(value, str):
        if RFC3339_PATTERN.fullmatch(value) is None:
            raise FGraphTypeError(f"invalid instant {value!r}; use RFC 3339 such as 2026-08-24T10:00:00Z")
        try:
            parsed = datetime.fromisoformat(value)
        except ValueError as exc:
            raise FGraphTypeError(f"invalid instant {value!r}; use RFC 3339 such as 2026-08-24T10:00:00Z") from exc
        if parsed.tzinfo is None:
            raise FGraphTypeError(f"instant {value!r} has no timezone; include Z or an explicit UTC offset")
        epoch = datetime(1970, 1, 1, tzinfo=UTC)
        try:
            delta = parsed.astimezone(UTC) - epoch
        except (OverflowError, ValueError) as exc:
            raise FGraphTypeError(
                f"instant {value!r} normalizes outside RFC 3339 years 0001..9999; use a representable UTC instant"
            ) from exc
        return (delta.days * 86_400 + delta.seconds) * 1_000_000 + delta.microseconds
    raise FGraphTypeError(f"invalid instant {value!r}; use integer microseconds or an RFC 3339 UTC string")


def _bytes(value: Any) -> bytes:
    if isinstance(value, bytes):
        result = value
    elif isinstance(value, str):
        try:
            result = base64.b64decode(value, validate=True)
        except (binascii.Error, ValueError) as exc:
            raise FGraphTypeError(f"invalid bytes value {value!r}; use standard padded base64") from exc
    else:
        raise FGraphTypeError(f"invalid bytes value {value!r}; use bytes or a standard padded base64 string")
    _check_size(result, value)
    return result


def _vector(value: Any) -> tuple[tuple[float, ...], bytes]:
    if not isinstance(value, (list, tuple)) or not value:
        raise FGraphTypeError(f"invalid vector {value!r}; use a non-empty array of finite numbers")
    converted: list[float] = []
    for item in value:
        if isinstance(item, bool) or not isinstance(item, (int, float)):
            raise FGraphTypeError(f"invalid vector element {item!r}; use only finite numbers")
        number = float(item)
        if not math.isfinite(number):
            raise FGraphTypeError(f"invalid vector element {item!r}; NaN and infinity are not supported")
        converted.append(number)
    try:
        packed = struct.pack(f"<{len(converted)}f", *converted)
    except (OverflowError, struct.error) as exc:
        raise FGraphTypeError(
            f"vector {value!r} cannot be represented as float32; reduce its element magnitudes"
        ) from exc
    _check_size(packed, value)
    # Reading the float32 bytes back makes every API observe the persisted precision.
    logical = struct.unpack(f"<{len(converted)}f", packed)
    return logical, packed


def encode(value: Any, resolve_ref: Callable[[Any], int] | None = None) -> Encoded:
    """Parse one Python/wire value into a tagged SQLite value."""
    if isinstance(value, dict):
        if len(value) != 1:
            raise FGraphTypeError(
                f"value object {value!r} is not a typed wrapper; wrap literal objects with {{'json': ...}}"
            )
        kind, inner = next(iter(value.items()))
        if kind == "ref":
            if resolve_ref is None:
                raise FGraphTypeError(f"reference {inner!r} cannot be resolved here; provide an entity id or name")
            target = resolve_ref(inner)
            return Encoded(REF, target, target)
        if kind == "instant":
            instant = _instant(inner)
            return Encoded(INSTANT, instant, instant)
        if kind == "bytes":
            data = _bytes(inner)
            if len(data) > BLOB_THRESHOLD:
                digest = indirect_digest(BYTES_REF, data)
                return Encoded(BYTES_REF, digest, data, data)
            return Encoded(BYTES, data, data)
        if kind == "vector":
            logical, packed = _vector(inner)
            digest = indirect_digest(VECTOR, packed)
            return Encoded(VECTOR, digest, logical, packed)
        if kind == "json":
            canonical = canonical_json(inner)
            _check_size(canonical.encode(), inner)
            return Encoded(JSON, canonical, canonical)
        raise FGraphTypeError(
            f"unknown typed wrapper {kind!r}; use ref, instant, bytes, vector, or json (tmp is only valid as an id)"
        )
    if value is None:
        raise FGraphTypeError("null is not a fact scalar; wrap it as {'json': null} when null is domain data")
    if isinstance(value, bool):
        return Encoded(BOOL, int(value), value)
    if isinstance(value, int):
        if not INT64_MIN <= value <= INT64_MAX:
            raise FGraphTypeError(f"integer {value!r} exceeds signed 64-bit range; store a smaller integer or text")
        return Encoded(INT, value, value)
    if isinstance(value, float):
        if not math.isfinite(value):
            raise FGraphTypeError(f"float {value!r} is not finite; NaN and infinity are not supported")
        return Encoded(FLOAT, value, value)
    if isinstance(value, str):
        data = _utf8(value)
        _check_size(data, value)
        if len(data) > BLOB_THRESHOLD:
            digest = indirect_digest(TEXT_REF, data)
            return Encoded(TEXT_REF, digest, value, value)
        return Encoded(TEXT, value, value)
    if isinstance(value, bytes):
        data = _bytes(value)
        if len(data) > BLOB_THRESHOLD:
            digest = indirect_digest(BYTES_REF, data)
            return Encoded(BYTES_REF, digest, data, data)
        return Encoded(BYTES, data, data)
    raise FGraphTypeError(f"unsupported fact value {value!r}; use a scalar or a typed wrapper")


def type_name(tag: int) -> str:
    """Return the logical schema type for a physical value tag."""
    return TAG_NAMES[tag]


def value_matches(declared: str | None, encoded: Encoded) -> bool:
    """Check a physical tag against an optional logical schema type."""
    return declared is None or type_name(encoded.tag) == declared


def instant_text(microseconds: int) -> str:
    """Render integer UTC microseconds as fixed-width RFC 3339."""
    microseconds = _instant(microseconds)
    seconds, micros = divmod(microseconds, 1_000_000)
    # Datetime arithmetic supports the full normative range on Windows too;
    # fromtimestamp delegates to a platform C runtime that may reject year 1.
    moment = _UNIX_EPOCH + timedelta(seconds=seconds)
    rendered = (
        f"{moment.year:04d}-{moment.month:02d}-{moment.day:02d}T"
        f"{moment.hour:02d}:{moment.minute:02d}:{moment.second:02d}"
    )
    return f"{rendered}.{micros:06d}Z"


def wire_value(tag: int, logical: Any, name_for_id: Callable[[int], Any]) -> Any:
    """Render a logical value in the normative JSON wire form."""
    if tag == REF:
        return {"ref": name_for_id(int(logical))}
    if tag == INSTANT:
        return {"instant": instant_text(int(logical))}
    if tag in (BYTES, BYTES_REF):
        return {"bytes": base64.b64encode(logical).decode("ascii")}
    if tag == VECTOR:
        return {"vector": list(logical)}
    if tag == JSON:
        return {"json": json.loads(logical)}
    return logical

File: typescript/src/values.ts
import { createHash } from "node:crypto";

import { TooLarge, TypeError } from "./errors.js";
import { JsonFloat, canonicalValueJson, parseJsonValue } from "./jsonio.js";

export const BLOB_THRESHOLD = 256;
export const MAX_VALUE_BYTES = 1_048_576;
export const INT64_MIN = -(2n ** 63n);
export const INT64_MAX = 2n ** 63n - 1n;
export const INSTANT_MIN = -62_135_596_800_000_000n;
export const INSTANT_MAX = 253_402_300_799_999_999n;
export const ATTRIBUTE_PATTERN =
  /^[a-z0-9][a-z0-9._-]*\/[a-z0-9][a-z0-9._-]*$/u;

export const REF = 0;
export const BOOL = 1;
export const INT = 2;
export const FLOAT = 3;
export const TEXT = 4;
export const INSTANT = 5;
export const BYTES = 6;
export const VECTOR = 7;
export const TEXT_REF = 8;
export const BYTES_REF = 9;
export const JSON_TAG = 10;

export type Stored = bigint | number | string | Buffer;
export interface Encoded {
  tag: number;
  stored: Stored;
  logical: unknown;
  blob?: string | Buffer;
}

export interface Cell {
  tag: number;
  value: unknown;
}

export const TYPE_NAMES = new Set([
  "ref",
  "bool",
  "int",
  "float",
  "text",
  "instant",
  "bytes",
  "vector",
  "json",
]);
const TAG_NAMES = [
  "ref",
  "bool",
  "int",
  "float",
  "text",
  "instant",
  "bytes",
  "vector",
  "text",
  "bytes",
  "json",
] as const;

function utf8(value: string): Buffer {
  if (!value.isWellFormed()) {
    throw new TypeError(
      `string ${JSON.stringify(value)} is not valid UTF-8; remove unpaired surrogate code points`,
    );
  }
  return Buffer.from(value, "utf8");
}

function checkSize(data: Uint8Array, value: unknown): void {
  if (data.byteLength > MAX_VALUE_BYTES) {
    throw new TooLarge(
      `value ${typeof value} is ${data.byteLength} bytes; keep fact values at or below ${MAX_VALUE_BYTES} bytes`,
    );
  }
}

export function indirectDigest(tag: number, data: Uint8Array): Buffer {
  // The physical tag keeps byte-identical values from different domains apart.
  return createHash("sha256")
    .update(Buffer.from([tag]))
    .update(data)
    .digest();
}

function asInt64(value: unknown): bigint {
  if (typeof value === "bigint") {
    if (value >= INT64_MIN && value <= INT64_MAX) return value;
  } else if (typeof value === "number" && Number.isSafeInteger(value)) {
    return BigInt(value);
  }
  throw new TypeError(
    `integer ${String(value)} exceeds signed 64-bit range; store a smaller integer or text`,
  );
}

const RFC3339 =
  /^([0-9]{4})-([0-9]{2})-([0-9]{2})T([0-9]{2}):([0-9]{2}):([0-9]{2})(?:\.([0-9]+))?(Z|[+-][0-9]{2}:[0-9]{2})$/u;

export function instantValue(value: unknown): bigint {
  if (typeof value === "number" || typeof value === "bigint") {
    const instant = asInt64(value);
    if (instant >= INSTANT_MIN && instant <= INSTANT_MAX) return instant;
    throw new TypeError(
      `instant ${String(value)} is outside RFC 3339 years 0001..9999; use representable UTC microseconds`,
    );
  }
  if (typeof value !== "string") {
    throw new TypeError(
      `invalid instant ${String(value)}; use integer microseconds or an RFC 3339 UTC string`,
    );
  }
  const match = RFC3339.exec(value);
  if (match === null)
    throw new TypeError(
      `invalid instant ${JSON.stringify(value)}; use RFC 3339 such as 2026-08-24T10:00:00Z`,
    );
  const [
    ,
    yearText,
    monthText,
    dayText,
    hourText,
    minuteText,
    secondText,
    fraction = "",
    zone,
  ] = match;
  const year = Number(yearText);
  const month = Number(monthText);
  const day = Number(dayText);
  const hour = Number(hourText);
  const minute = Number(minuteText);
  const second = Number(secondText);
  const fractionMicros = BigInt((fraction + "000000").slice(0, 6));
  const base = new Date(0);
  base.setUTCFullYear(year, month - 1, day);
  base.setUTCHours(hour, minute, second, 0);
  if (
    base.getUTCFullYear() !== year ||
    base.getUTCMonth() !== month - 1 ||
    base.getUTCDate() !== day ||
    base.getUTCHours() !== hour ||
    base.getUTCMinutes() !== minute ||
    base.getUTCSeconds() !== second
  ) {
    throw new TypeError(
      `invalid instant ${JSON.stringify(value)}; use RFC 3339 such as 2026-08-24T10:00:00Z`,
    );
  }
  let offsetMinutes = 0;
  if (zone !== "Z") {
    const sign = zone?.[0] === "+" ? 1 : -1;
    const zoneHour = Number(zone?.slice(1, 3));
    const zoneMinute = Number(zone?.slice(4, 6));
    if (zoneHour > 23 || zoneMinute > 59)
      throw new TypeError(
        `invalid instant ${JSON.stringify(value)}; use a valid UTC offset`,
      );
    offsetMinutes = sign * (zoneHour * 60 + zoneMinute);
  }
  const result =
    BigInt(base.getTime() - offsetMinutes * 60_000) * 1000n + fractionMicros;
  if (result < INSTANT_MIN || result > INSTANT_MAX)
    throw new TypeError(
      `instant ${JSON.stringify(value)} is outside RFC 3339 years 0001..9999`,
    );
  return result;
}

function bytesValue(value: unknown): Buffer {
  let result: Buffer;
  if (Buffer.isBuffer(value)) result = Buffer.from(value);
  else if (value instanceof Uint8Array) result = Buffer.from(value);
  else if (typeof value === "string") {
    result = Buffer.from(value, "base64");
    if (result.toString("base64") !== value)
      throw new TypeError(
        `invalid bytes value ${JSON.stringify(value)}; use standard padded base64`,
      );
  } else
    throw new TypeError(
      `invalid bytes value ${String(value)}; use bytes or a standard padded base64 string`,
    );
  checkSize(result, value);
  return result;
}

function vectorValue(value: unknown): { logical: number[]; packed: Buffer } {
  if (!Array.isArray(value) || value.length === 0)
    throw new TypeError(
      `invalid vector ${String(value)}; use a non-empty array of finite numbers`,
    );
  const packed = Buffer.allocUnsafe(value.length * 4);
  const logical: number[] = [];
  for (const [index, item] of value.entries()) {
    if (
      (typeof item !== "number" &&
        typeof item !== "bigint" &&
        !(item instanceof JsonFloat)) ||
      !Number.isFinite(Number(item))
    ) {
      throw new TypeError(
        `invalid vector element ${String(item)}; use only finite numbers`,
      );
    }
    packed.writeFloatLE(Number(item), index * 4);
    const rounded = packed.readFloatLE(index * 4);
    if (!Number.isFinite(rounded))
      throw new TypeError(
        `vector element ${String(item)} cannot be represented as float32; reduce its magnitude`,
      );
    logical.push(rounded);
  }
  checkSize(packed, value);
  return { logical, packed };
}

export function encode(
  value: unknown,
  resolveRef?: (value: unknown) => bigint,
): Encoded {
  if (isRecord(value)) {
    const entries = Object.entries(value);
    if (entries.length !== 1)
      throw new TypeError(
        `value object is not a typed wrapper; wrap literal objects with {"json":...}`,
      );
    const [kind, inner] = entries[0] ?? [];
    if (kind === "ref") {
      if (resolveRef === undefined)
        throw new TypeError(
          `reference ${String(inner)} cannot be resolved here; provide an entity id or name`,
        );
      const target = resolveRef(inner);
      return { tag: REF, stored: target, logical: target };
    }
    if (kind === "instant") {
      const instant = instantValue(inner);
      return { tag: INSTANT, stored: instant, logical: instant };
    }
    if (kind === "bytes") {
      const data = bytesValue(inner);
      return data.length > BLOB_THRESHOLD
        ? {
            tag: BYTES_REF,
            stored: indirectDigest(BYTES_REF, data),
            logical: data,
            blob: data,
          }
        : { tag: BYTES, stored: data, logical: data };
    }
    if (kind === "vector") {
      const { logical, packed } = vectorValue(inner);
      return {
        tag: VECTOR,
        stored: indirectDigest(VECTOR, packed),
        logical,
        blob: packed,
      };
    }
    if (kind === "json") {
      const canonical = canonicalValueJson(inner);
      checkSize(Buffer.from(canonical), inner);
      return { tag: JSON_TAG, stored: canonical, logical: canonical };
    }
    throw new TypeError(
      `unknown typed wrapper ${String(kind)}; use ref, instant, bytes, vector, or json (tmp is only valid as an id)`,
    );
  }
  if (value === null || value === undefined)
    throw new TypeError(
      `null is not a fact scalar; wrap it as {"json":null} when null is domain data`,
    );
  if (typeof value === "boolean")
    return { tag: BOOL, stored: value ? 1n : 0n, logical: value };
  if (value instanceof JsonFloat) {
    if (!Number.isFinite(value.value))
      throw new TypeError(
        `float ${String(value)} is not finite; NaN and infinity are not supported`,
      );
    return { tag: FLOAT, stored: value.value, logical: value.value };
  }
  if (typeof value === "bigint") {
    const integer = asInt64(value);
    return { tag: INT, stored: integer, logical: integer };
  }
  if (typeof value === "number") {
    if (!Number.isFinite(value))
      throw new TypeError(
        `float ${String(value)} is not finite; NaN and infinity are not supported`,
      );
    if (Number.isInteger(value)) {
      const integer = asInt64(value);
      return { tag: INT, stored: integer, logical: integer };
    }
    return { tag: FLOAT, stored: value, logical: value };
  }
  if (typeof value === "string") {
    const data = utf8(value);
    checkSize(data, value);
    return data.length > BLOB_THRESHOLD
      ? {
          tag: TEXT_REF,
          stored: indirectDigest(TEXT_REF, data),
          logical: value,
          blob: value,
        }
      : { tag: TEXT, stored: value, logical: value };
  }
  if (Buffer.isBuffer(value) || value instanceof Uint8Array) {
    const data = bytesValue(value);
    return data.length > BLOB_THRESHOLD
      ? {
          tag: BYTES_REF,
          stored: indirectDigest(BYTES_REF, data),
          logical: data,
          blob: data,
        }
      : { tag: BYTES, stored: data, logical: data };
  }
  throw new TypeError(
    `unsupported fact value ${String(value)}; use a scalar or a typed wrapper`,
  );
}

export function typeName(tag: number): string {
  const name = TAG_NAMES[tag];
  if (name === undefined)
    throw new TypeError(
      `unknown stored value tag ${tag}; run doctor to inspect file corruption`,
    );
  return name;
}

export function valueMatches(declared: string | null, value: Encoded): boolean {
  return declared === null || typeName(value.tag) === declared;
}

export function instantText(microseconds: bigint): string {
  const instant = instantValue(microseconds);
  let seconds = instant / 1_000_000n;
  let micros = instant % 1_000_000n;
  if (micros < 0) {
    seconds -= 1n;
    micros += 1_000_000n;
  }
  const iso = new Date(Number(seconds * 1000n)).toISOString();
  return `${iso.slice(0, 19)}.${micros.toString().padStart(6, "0")}Z`;
}

export function wireValue(
  tag: number,
  logical: unknown,
  nameForId: (id: bigint) => string | number | bigint,
): unknown {
  if (tag === REF) return { ref: nameForId(logical as bigint) };
  if (tag === INSTANT) return { instant: instantText(logical as bigint) };
  if (tag === BYTES || tag === BYTES_REF)
    return { bytes: (logical as Buffer).toString("base64") };
  if (tag === VECTOR) return { vector: logical };
  if (tag === JSON_TAG)
    return {
      json: materializeJson(parseJsonValue(logical as string, "stored JSON")),
    };
  if (tag === INT) return publicInteger(logical as bigint);
  return logical;
}

function materializeJson(value: unknown): unknown {
  if (value instanceof JsonFloat) return value.value;
  if (Array.isArray(value)) return value.map(materializeJson);
  if (value !== null && typeof value === "object")
    return Object.fromEntries(
      Object.entries(value).map(([key, item]) => [key, materializeJson(item)]),
    );
  return value;
}

export function publicInteger(value: bigint): number | bigint {
  return value >= BigInt(Number.MIN_SAFE_INTEGER) &&
    value <= BigInt(Number.MAX_SAFE_INTEGER)
    ? Number(value)
    : value;
}

export function isRecord(value: unknown): value is Record<string, unknown> {
  return (
    typeof value === "object" &&
    value !== null &&
    !Array.isArray(value) &&
    !ArrayBuffer.isView(value) &&
    !(value instanceof JsonFloat)
  );
}

export function equalStored(left: Stored, right: Stored): boolean {
  return Buffer.isBuffer(left) && Buffer.isBuffer(right)
    ? left.equals(right)
    : left === right;
}

export function storedKey(value: Stored): string {
  if (Buffer.isBuffer(value)) return `b:${value.toString("hex")}`;
  return `${typeof value}:${String(value)}`;
}
